From 6bb311ed60527ff895b386cc86393d6e83fb4557 Mon Sep 17 00:00:00 2001
From: Gregor Riepl <onitake@gmail.com>
Date: Tue, 24 May 2016 08:15:52 +0200
Subject: [PATCH 2/2] Improve patch for #134: add NULL checks

---
 src/fserve.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/fserve.c b/src/fserve.c
index 5207c1e..47069c5 100644
--- a/src/fserve.c
+++ b/src/fserve.c
@@ -972,12 +972,16 @@ int fserve_setup_client_fb (client_t *client, fbinfo *finfo)
     if (client->check_buffer == NULL)
         client->check_buffer = format_generic_write_to_client;
 
+    // workaround for #134: fill the preallocated, but empty, chained buffer in case a range request was made
     if (client->flags & CLIENT_RANGE_END)
     {
-        refbuf = client->refbuf->next;
-        bytes = pread (fh->f, refbuf->data, refbuf->len, client->intro_offset);
-        if (bytes < 0)
-            return -1;
+        if (client->refbuf && client->refbuf->next)
+        {
+            refbuf = client->refbuf->next;
+            bytes = pread (fh->f, refbuf->data, refbuf->len, client->intro_offset);
+            if (bytes < 0)
+                return -1;
+        }
     }
     
     client->ops = &buffer_content_ops;
-- 
2.7.4

