diff --git a/src/fserve.c b/src/fserve.c
index 276f07b..5207c1e 100644
--- a/src/fserve.c
+++ b/src/fserve.c
@@ -883,6 +883,8 @@ int fserve_setup_client_fb (client_t *client, fbinfo *finfo)
 {
     fh_node *fh = &no_file;
     int ret = 0;
+    refbuf_t *refbuf;
+    ssize_t bytes;
 
     if (finfo)
     {
@@ -970,6 +972,14 @@ int fserve_setup_client_fb (client_t *client, fbinfo *finfo)
     if (client->check_buffer == NULL)
         client->check_buffer = format_generic_write_to_client;
 
+    if (client->flags & CLIENT_RANGE_END)
+    {
+        refbuf = client->refbuf->next;
+        bytes = pread (fh->f, refbuf->data, refbuf->len, client->intro_offset);
+        if (bytes < 0)
+            return -1;
+    }
+    
     client->ops = &buffer_content_ops;
     client->flags &= ~CLIENT_HAS_INTRO_CONTENT;
     client->flags |= CLIENT_IN_FSERVE;
